pandas/core/indexing.py

```diff
@@ -1767,15 +1768,20 @@ def _setitem_with_indexer_frame_value(self, indexer, value: "DataFrame"):
             raise ValueError("Setting with non-unique columns is not allowed.")
 
         else:
+            index = 0
             for loc in ilocs:
                 item = self.obj.columns[loc]
                 if item in value:
                     sub_indexer[1] = item
-                    val = self._align_series(
-                        tuple(sub_indexer), value[item], multiindex_indexer
-                    )
+                    if self.name == "loc":
+                        val = self._align_series(
+                            tuple(sub_indexer), value[item], multiindex_indexer
+                        )
+                    else:
+                        val = value.iloc[:, index]
```

Comment:
could you do the self.name check (or whatever that check turns out to look like) outside of the loop?

Reply:
Thanks, added this check before the if conditions, this is much better.